lib/HalDependencyResolver.js

```diff
@@ -400,54 +403,58 @@ HalDependencyResolver.prototype = {
-				//	  |        \\       /  \       /  \       /         (_I_)   (_{;}_)|   | \ (_ o _) /|  |    |  |
-				//	  `--------` `'-..-'    `-...-'    `-...-'          '---'   '(_,_) '---'  '.(_,_).' '--'    '--'
-				//
-
-				//arr.push(m);
-
-				// instead of returning the module we found, lets return the module with the version we need,
-				// and any dependencies it requires, I think that's more clear.
-
-				var missing = extend(m, { v: needs.v });
-				arr.push(missing);
-				// todo mdm - this is wrong. we shouldn't be fabricating dependencies for a new version of a module
-				// (need) from the installed version (m) - dependencies can and do change between versions.
-				// The database of known modules should be the definitive source of module dependencies.
-
-				// if we're updating this, we better check its dependencies too.
-				if (m.d && (m.d.length > 0)) {
-					//oh no!  this module has dependencies!
-					// do we have to update those too?
-					// (this doesn't fully make sense to me right now, but lets go with it.)
-					// todo mdm - this won't do anything, since `m` is always a satisfied dependency
-					// (it came from the modules array.)
-					// at the very least we should be iterating over m.d[] as the needed modules
-					arr = this._walkChain(modules, m, arr);
-				}
+		if (!foundModule) {
+			// Module is not present in the list of modules on the device
+			// Simply return the dependency as-is
+			arr.push(needs);
+		} else if (foundModule.v < needs.v) {
+			//
+			// it's...
+			//	  .---.       .-''-.     .-'''-.    .-'''-.     ,---------. .---.  .---.    ____    ,---.   .--.
+			//	  | ,_|     .'_ _   \   / _     \  / _     \    \          \|   |  |_ _|  .'  __ `. |    \  |  |
+			//	,-./  )    / ( ` )   ' (`' )/`--' (`' )/`--'     `--.  ,---'|   |  ( ' ) /   '  \  \|  ,  \ |  |
+			//	\  '_ '`) . (_ o _)  |(_ o _).   (_ o _).           |   \   |   '-(_{;}_)|___|  /  ||  |\_ \|  |
+			//	 > (_)  ) |  (_,_)___| (_,_). '.  (_,_). '.         :_ _:   |      (_,_)    _.-`   ||  _( )_\  |
+			//	(  .  .-' '  \   .---..---.  \  :.---.  \  :        (_I_)   | _ _--.   | .'   _    || (_ o _)  |
+			//	 `-'`-'|___\  `-'    /\    `-'  |\    `-'  |       (_(=)_)  |( ' ) |   | |  _( )_  ||  (_,_)\  |
+			//	  |        \\       /  \       /  \       /         (_I_)   (_{;}_)|   | \ (_ o _) /|  |    |  |
+			//	  `--------` `'-..-'    `-...-'    `-...-'          '---'   '(_,_) '---'  '.(_,_).' '--'    '--'
+			//
+
+			//arr.push(foundModule);
+
+			// instead of returning the module we found, lets return the module with the version we need,
+			// and any dependencies it requires, I think that's more clear.
+
+			var missing = extend(foundModule, { v: needs.v });
+			arr.push(missing);
+			// todo mdm - this is wrong. we shouldn't be fabricating dependencies for a new version of a module
+			// (need) from the installed version (m) - dependencies can and do change between versions.
```

Comment:
Let's rename `m` to `foundModule` in these comments as well (or vice versa), otherwise the comments don't make any sense anymore, and they're already confusing